Clear the transaction state when AR object is duped

yoreek opened the pull request:
Fixes #31670

yoreek committed:
Clear the transaction state when AR object is duped

kamipo commented:
Original issue is "ActiveRecord persisted? returns true after rollback when using duped record".

Can you test that issue is fixed?

yoreek commented:
@kamipo, here is the test before and after the fix:

$ git co master 
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
$ bundle exec ../persisted_after_rollback.rb
Active Record 5.2.0.beta2
-- create_table(:users, {:force=>true})
   -> 0.0050s
Create a clone using dup
persisted? = true
Create a clone using new
persisted? = false

$ git co clear_transaction_state_when_ar_object_is_duped
Switched to branch 'clear_transaction_state_when_ar_object_is_duped'
$ bundle exec ../persisted_after_rollback.rb            
Active Record 5.2.0.beta2
-- create_table(:users, {:force=>true})
   -> 0.0050s
Create a clone using dup
persisted? = false
Create a clone using new
persisted? = false

kamipo commented:
Yeah, can you add the test in the commit to catch a future regression?

yoreek commented:
@kamipo done

kamipo commented on activerecord/test/cases/dup_test.rb:
```diff
@@ -157,5 +158,34 @@ def test_dup_without_primary_key
+    def test_dup_transaction_state_and_transaction_record_state_are_cleared
+      topic = Topic.first
+      topic.send(:set_transaction_state, ActiveRecord::ConnectionAdapters::TransactionState.new(:rollback))
+      topic.send(:remember_transaction_record_state)
```
We don't need to test internal states with private methods as long as it can be tested from public API.
And also, this test case does not show that the state of the record when the transaction was rolled back has been fixed.

yoreek replied:
Thanks for review.
This test only checks that the transaction state is cleared in the duped record. I think that we must clear this state as well as other statuses. Do you think it is necessary to test the complete scheme with the rollback of the transaction?
And, unfortunately I did not find a public methods to access to `@transaction_state` and `@_start_transaction_state`, could you help me ?

kamipo replied:
Since these are internal states we don't have any way to access to the states.
So it is enough to test the transaction behavior which tested in the following test rather than testing the states.

The `test_dup_record_not_persisted_after_rollback_transaction` sufficiently shows which behavior is fixed.
So we simply remove the `test_dup_transaction_state_and_transaction_record_state_are_cleared`.

yoreek replied:
ok, I removed the test and added the entry to CHANGELOG.

kamipo reviewed (changes requested):
Can you add CHANGELOG entry what issue is fixed?
